jgrapht-core/src/main/java/org/jgrapht/alg/lca/BinaryLiftingLCAFinder.java

```diff
@@ -0,0 +1,204 @@
+     * @param root the root of the graph
+     */
+    public BinaryLiftingLCAFinder(Graph<V, E> graph, V root){
+        this(graph, Collections.singleton(Objects.requireNonNull(root, "root cannot be null")));
+    }
+
+    /**
+     * Construct a new instance of the algorithm.
+     *
+     * Note: If two roots appear in the same tree, an error will be thrown.
+     * Note: The constructor will NOT check if the input graph is a valid forest.
+     *
+     * @param graph the input graph
+     * @param roots the set of roots of the graph
+     */
+    public BinaryLiftingLCAFinder(Graph<V, E> graph, Set<V> roots){
+        this.graph = Objects.requireNonNull(graph, "graph cannot be null");
+        this.roots = Objects.requireNonNull(roots, "roots cannot be null");
+        this.MAX_LEVEL = log2(graph.vertexSet().size());
+
+        if (this.roots.isEmpty())
+            throw new IllegalArgumentException("roots cannot be empty");
+
+        if (!graph.vertexSet().containsAll(roots))
+            throw new IllegalArgumentException("at least one root is not a valid vertex");
+
+        computeAncestorMatrix();
+    }
+
+    private void normalizeGraph(){
+        /*
+         * Normalize the graph map each vertex to an integer (using a HashMap) keep the reverse
+         * mapping (using an ArrayList)
+         */
+        vertexMap = new HashMap<>(graph.vertexSet().size());
+        indexList = new ArrayList<>(graph.vertexSet().size());
+
+        for (V v : graph.vertexSet()) {
+            if (!vertexMap.containsKey(v)) {
+                vertexMap.put(v, vertexMap.size());
+                indexList.add(v);
+            }
+        }
+    }
+
+    private void dfs(int u, int parent){
+        component[u] = numberComponent;
+        timeIn[u] = ++clock;
+
+        ancestors[0][u] = parent;
+        for (int l = 1; l < MAX_LEVEL; l++) {
+            if (ancestors[l - 1][u] != -1)
+                ancestors[l][u] = ancestors[l - 1][ancestors[l - 1][u]];
+        }
+
+        V vertexU = indexList.get(u);
+        for (E edge: graph.edgesOf(vertexU)){
+            int v = vertexMap.get(Graphs.getOppositeVertex(graph, edge, vertexU));
+
+            if (v != parent){
+                dfs(v, u);
+            }
+        }
+
+        timeOut[u] = ++clock;
+    }
+
+    private void computeAncestorMatrix(){
+        ancestors = new int[MAX_LEVEL + 1][graph.vertexSet().size()];
+
+        for (int l = 0; l < MAX_LEVEL; l++) {
+            Arrays.fill(ancestors[l], -1);
+        }
+
+        timeIn = new int[graph.vertexSet().size()];
+        timeOut = new int[graph.vertexSet().size()];
+
+        // Ensure that isAncestor(x, y) == false if either x and y hasn't been explored yet
+        for (int i = 0; i < graph.vertexSet().size(); i++) {
+            timeIn[i] = timeOut[i] = -(i + 1);
+        }
+
+        numberComponent = 0;
+        component = new int[graph.vertexSet().size()];
+
+        normalizeGraph();
+
+        for (V root: roots)
```

Comment:
I know a few people like this, but it looks like Python :-) (not that I have something against Python)!

 Please add curly braces.